ibridges/irodsconnector/sync.py

```diff
@@ -0,0 +1,377 @@
+                    name=file,
+                    path=fix_local_path(rel_path),
+                    size=full_path.stat().st_size,
+                    checksum=None if ignore_checksum else _calc_checksum(full_path)))
+
+        collections.extend([FolderObject(
+                fix_local_path(str(Path(root) / dir)[len(str(path)):].lstrip(os.sep)),
+                len([x for x in Path(f"{root}{os.sep}{dir}").iterdir() if x.is_file()]),
+                len([x for x in Path(f"{root}{os.sep}{dir}").iterdir() if x.is_dir()])
+            )
+            for dir in dirs if max_level is None or dir.count(os.sep)<max_level-1])
+
+    return objects, collections
+
+def _get_irods_tree(coll: iRODSCollection,
+                    root: str = '',
+                    level: int = 0,
+                    max_level: Union[int, None] = None,
+                    ignore_checksum: bool = False):
+
+    root=coll.path if len(root)==0 else root
+
+    # chksum() (re)calculates checksum, call only when checksum is empty
+    objects=[FileObject(
+        x.name,
+        x.path[len(root):].lstrip('/'),
+        x.size,
+        None if ignore_checksum else (x.checksum if len(x.checksum)>0 else x.chksum()))
+        for x in coll.data_objects]
+
+    if max_level is None or level<max_level-1:
+        collections=[FolderObject(
+            x.path[len(root):].lstrip('/'),
+            len(x.data_objects),
+            len(x.subcollections)) for x in coll.subcollections]
+
+        for subcoll in coll.subcollections:
+            subobjects, subcollections=_get_irods_tree(
+                coll=subcoll,
+                root=root,
+                level=level+1,
+                max_level=max_level,
+                ignore_checksum=ignore_checksum
+                )
+            objects.extend(subobjects)
+            collections.extend(subcollections)
+    else:
+        collections=[]
+
+    return objects, collections
+
+def _create_irods_collections(session: Session,
+                              target: IrodsPath,
+                              collections: list[FolderObject],
+                              dry_run: bool,
+                              copy_empty_folders: bool):
+    new_colls=[str(target / x.path) for x in collections if not x.is_empty() or copy_empty_folders]
+    if dry_run:
+        print("Will create collection(s):")
+        print(*new_colls, sep='\n')
+        return
+    [create_collection(session, x) for x in new_colls]
+
+def _create_local_folders(target: Path,
+                          folders: list[FolderObject],
+                          dry_run: bool,
+                          copy_empty_folders: bool):
+    new_folders=[target / Path(x.path) for x in folders if not x.is_empty() or copy_empty_folders]
+    if dry_run:
+        print("Will create folder(s):")
+        print(*[f"  {x}" for x in new_folders], sep='\n')
+        return
+    [x.mkdir(parents=True, exist_ok=True) for x in new_folders]
```

Comment:
If you're not using the list in the comprehension, I think a simple for loop is better (even though it's one more line).